### PR TITLE
UMD wrapper for twine

### PR DESCRIPTION
**Pull request by tsov**
```js
import Twine from './twine';
```

❤️ enough said

@Shopify/tnt

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -9,7 +9,7 @@ TEST_FILES=$(shell find test/*.coffee)
 .pretestem: .all
 
 .app: $(APP_FILES)
-	node_modules/.bin/browserify $(APP_FILES) -o dist/twine.js -t coffeeify
```
Comment (lemonmade) [inline Makefile]: Was there no reason to use browserify before? Might want to use/ ditch `APP_FILES`

Comment (tsov) [inline Makefile]: Yeah, it was a little weird, because Twine got wrapped twice before. Yes, `APP_FILES` can 🔥

File: dist/twine.js [hunk 2]
```diff
@@ -1,537 +1,510 @@
...
-  if (node == null) {
-    node = document.documentElement;
-  }
-  for (key in elements) {
-    if (bindings = (ref = elements[key]) != null ? ref.bindings : void 0) {
-      for (j = 0, len = bindings.length; j < len; j++) {
-        obj = bindings[j];
-        if (obj.teardown) {
-          obj.teardown();
+// Generated by CoffeeScript 1.10.0
```
Comment (tylerball) [inline dist/twine.js]: is there a setting to remove this comment?

File: package.json [hunk 3]
```diff
@@ -30,7 +30,7 @@
     "coffeeify": "~0.7.0",
     "uglify-js": "~2.4.15",
     "browserify": "5.11.1",
-    "pbind": "~0.0.1",
```
Comment (tsov) [inline package.json]: 🔥 

Not used at all

File: lib/assets/javascripts/twine.coffee [hunk 4]
```diff
@@ -0,0 +1,349 @@
+((root, factory) ->
+  if typeof root.define == 'function' && root.define.amd
+    root.define
+  else if typeof module == 'object' && module.exports
+    module.exports = factory()
+  else
+    root.Twine = factory()
+) this, ->
```
Comment (lemonmade) [inline lib/assets/javascripts/twine.coffee]: Can you wrap the arguments in parens? I find it hard to read this in CoffeeScript

File: lib/assets/javascripts/twine.coffee [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+  Twine = {}
```
Comment (lemonmade) [inline lib/assets/javascripts/twine.coffee]: I assume you didn't change anything other than the wrapper?

File: dist/twine.js (hunk 2 above)
Comment (tsov) [inline dist/twine.js]: Ideally yes, but there is no option for `coffee` cli. Will figure something out, because it's annoying me too

**Comment by lemonmade** [comment]
Minor comments, otherwise looks awesome ❤️

File: lib/assets/javascripts/twine.coffee (hunk 4 above)
Comment (tsov) [inline lib/assets/javascripts/twine.coffee]: Sure 👍

File: lib/assets/javascripts/twine.coffee (hunk 5 above)
Comment (tsov) [inline lib/assets/javascripts/twine.coffee]: exactly, just `window.Twine` -> `Twine`

File: Makefile [hunk 6]
```diff
@@ -8,8 +7,8 @@ TEST_FILES=$(shell find test/*.coffee)
 # for now, exactly the same as .all
 .pretestem: .all
 
-.app: $(APP_FILES)
-	node_modules/.bin/browserify $(APP_FILES) -o dist/twine.js -t coffeeify
+.app:
+	node_modules/.bin/coffee -p lib/assets/javascripts/twine.coffee > dist/twine.js
```
Comment (tsov) [inline Makefile]: @tylerball only way to get rid of the comment.

Seems like all the requests to make that comment optional were declined... https://github.com/jashkenas/coffeescript/pull/3166

**Comment by tsov** [comment]
Good to 🐑  ?

File: Makefile (hunk 6 above)
Comment (tylerball) [inline Makefile]: That issue got closed with a commit that looks like an option got added: https://github.com/jashkenas/coffeescript/commit/2b03fa9077cea31b25cb23518d8af59fe5b80fed

Comment (tsov) [inline Makefile]: oups, didn't see it !! Thanks for that

File: Makefile [hunk 7]
```diff
@@ -8,8 +7,8 @@ TEST_FILES=$(shell find test/*.coffee)
 # for now, exactly the same as .all
 .pretestem: .all
 
-.app: $(APP_FILES)
-	node_modules/.bin/browserify $(APP_FILES) -o dist/twine.js -t coffeeify
+.app:
+	node_modules/.bin/coffee --no-header -o dist/ -c lib/assets/javascripts/twine.coffee
```
Comment (tsov) [inline Makefile]: @tylerball 👌

**Comment by tsov** [comment]
Ok, ran `make .uglify` too and added the minified version in this PR

**Comment by tylerball** [comment]
Nice. 👍

File: dist/twine.js [hunk 8]
```diff
@@ -1,537 +1,509 @@
...
-      for (j = 0, len = bindings.length; j < len; j++) {
-        obj = bindings[j];
-        if (obj.teardown) {
-          obj.teardown();
+(function() {
+  var slice = [].slice;
+
+  (function(root, factory) {
+    if (typeof root.define === 'function' && root.define.amd) {
+      return root.define;
```
Comment (owilliams) [inline dist/twine.js]: Is this correct for AMD? The [pattern defined on the github repo](https://github.com/umdjs/umd/blob/master/templates/returnExports.js#L44) seems to call `define([], factory)`

Comment (tsov) [inline dist/twine.js]: Nice catch, not sure how that happened. Will fix up 👌

**Comment by tsov** [comment]
@owilliams Updated

**Comment by owilliams** [comment]
👍

**Comment by qq99** [comment]
👏

**Pull request merged**